udp_test.go

```diff
@@ -0,0 +1,74 @@
+// Copyright (C) 2020 Storj Labs, Inc.
+// Copyright (C) 2014 Space Monkey, Inc.
+// See LICENSE for copying information.
+
+package jaeger
+
+import (
+	"testing"
+	"time"
+
+	"github.com/spacemonkeygo/monkit/v3"
+	"github.com/stretchr/testify/require"
+
+	"storj.io/monkit-jaeger/gen-go/jaeger"
+)
+
+func TestUDPCollector(t *testing.T) {
+	agentAdrr := "127.0.0.1:0"
+	mock, err := StartMockAgent(agentAdrr)
+	require.NoError(t, err)
+	defer func() {
+		err := mock.Close()
+		require.NoError(t, err)
+	}()
+	collector, err := NewUDPCollector(mock.Addr(), 200, "test", nil)
+	require.NoError(t, err)
+
+	span := &jaeger.Span{
+		TraceIdLow:    monkit.NewId(),
+		SpanId:        monkit.NewId(),
+		OperationName: "test-udp-collector",
+		StartTime:     time.Now().UnixNano() / 1000,
+		Duration:      time.Second.Microseconds(),
+	}
+	collector.Collect(span)
+
+	data := mock.GetBatches()
+	for len(data) < 1 {
+		time.Sleep(1 * time.Millisecond)
+		data = mock.GetBatches()
+	}
+	require.Len(t, data, 1)
+	require.Len(t, data[0].GetSpans(), 1)
+	receivedSpan := data[0].GetSpans()[0]
+	require.Equal(t, span.GetOperationName(), receivedSpan.OperationName)
+	require.Equal(t, span.GetTraceIdLow(), receivedSpan.GetTraceIdLow())
+	require.Equal(t, span.GetSpanId(), receivedSpan.GetSpanId())
+}
+
+func TestUDPCollector_HugeSpan(t *testing.T) {
+	agentAdrr := "127.0.0.1:0"
+	mock, err := StartMockAgent(agentAdrr)
+	require.NoError(t, err)
+	defer func() {
+		err := mock.Close()
+		require.NoError(t, err)
+	}()
+	collector, err := NewUDPCollector(mock.Addr(), 50, "test", nil)
+	require.NoError(t, err)
+
+	span := &jaeger.Span{
+		TraceIdLow:    monkit.NewId(),
+		SpanId:        monkit.NewId(),
+		OperationName: "test-udp-collector",
+		StartTime:     time.Now().UnixNano() / 1000,
+		Duration:      time.Second.Microseconds(),
+	}
+
+	collector.Collect(span)
+
+	time.Sleep(1 * time.Millisecond)
+	data := mock.GetBatches()
+	require.Len(t, data, 0)
```

Comment:
What is it you're testing here? I simply see that you're waiting a millisecond and verifying that it's 0. What if the only reason it passes is because the data hasn't arrived yet? This test can't guarantee that the data isn't arriving after the 1 millisecond.

We need a different way to test what you're looking here, I think.

Reply:
I think the test would be a lot easier to write if we exposed the `send()` method from the collector, but I don't know if it's good to make a method public just for testing?

Reply:
That depends. Sometimes it's fine to expose for testing only. Sometimes if it's really needed for a test that shows that it really should be a public method.

Sometimes it requires a different way of thinking about the test.

Either way, right now you're only guaranteeing that the data does not appear within 1 millisecond (that would be true no matter how long you slept), but you aren't guaranteeing that it will never appear, meaning this test is easy to break in the future.